app/src/main/java/io/gnosis/safe/ui/safe/terms/TermsViewModel.kt

```diff
@@ -0,0 +1,31 @@
+package io.gnosis.safe.ui.safe.terms
+
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import javax.inject.Inject
+
+class TermsViewModel @Inject constructor(
+    private val termsChecker: TermsChecker
+) : ViewModel() {
+
+    lateinit var advance: () -> Unit
+
+    val show = MutableLiveData<Boolean>()
+    private fun showTermsBottomSheet() {
+        show.postValue(true)
+    }
+
+    fun checkTerms(advance: () -> Unit) {
+        if (termsChecker.getTermsAgreed()) {
+            advance()
+        } else {
+            this.advance = advance
+            showTermsBottomSheet()
+        }
+    }
+
+    fun onAgreeClicked() {
+        termsChecker.setTermsAgreed(true)
```

Comment:
Here you could use your mutableLive data to communicate that the terms have been agreed to and remove the need to store a reference to `advance`